File: Theory_Of_Information/6 semester/laboratory__3/src/main/java/org/example/laboratory__3/coding/BlockCoder.java
package org.example.laboratory__3.coding;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Класс для реализации блочного кодирования
 */
public class BlockCoder {
    private List<String> codeWords;    // Список кодовых слов
    private List<String> informWords;
    private List<String> matrHT;
    private Map<String, String> sindromVector;
    private int n;                      // Длина кодового слова
    private int k;                      // Длина информационного слова
    private int dmin;                   // Минимальное расстояние
    private int r;                      // Количество ошибок, которые можно обнаружить
    private int t;                      // Корректирующая способность

    /**
     * Конструктор класса
     */
    public BlockCoder() {
        codeWords = new ArrayList<>();
        informWords = new ArrayList<>();
        matrHT = new ArrayList<>();
        sindromVector = new HashMap<>();
        n = 0;
        k = 0;
        dmin = 0;
        r = 0;
        t = 0;
    }

    /**
     * Logical XOR for an array of strings
     */
    public String xxor(List<String> array) {
        if (array == null || array.isEmpty()) {
            return "";
        }
        
        int[] result = new int[array.get(0).length()];
        for (String word : array) {
            for (int i = 0; i < word.length(); i++) {
                result[i] ^= Character.getNumericValue(word.charAt(i));
            }
        }
        
        StringBuilder sb = new StringBuilder();
        for (int bit : result) {
            sb.append(bit);
        }
        
        return sb.toString();
    }

    /**
     * Transpose a matrix
     */
    public List<String> transp(List<String> matrix) {
        if (matrix == null || matrix.isEmpty()) {
            return new ArrayList<>();
        }
        
        List<String> result = new ArrayList<>();
        for (int i = 0; i < matrix.get(0).length(); i++) {
            StringBuilder newRow = new StringBuilder();
            for (String row : matrix) {
                newRow.append(row.charAt(i));
            }
            result.add(newRow.toString());
        }
        
        return result;
    }

    /**
     * Generate matrices G and H based on input matrix
     */
    public List<String>[] generateMatrices(List<String> inputMatrix, String matrixType) {
        if (inputMatrix == null || inputMatrix.isEmpty()) {
            return null;
        }
        
        List<String> matrG = new ArrayList<>();
        List<String> matrH = new ArrayList<>();
        
        if ("H".equals(matrixType)) {
            // For special matrix H (4x7) we need to get n=9, k=5, dmin=2
            int rows = inputMatrix.size();     // 4 for example matrix
            int cols = inputMatrix.get(0).length();  // 7 for example matrix
            
            // Set known parameters for code n=9, k=5 for matrix H (4x7)
            int nTarget = 9;  // Desired code word length
            int kTarget = 5;  // Desired number of information bits
            
            // Create matrix H (extended version of input matrix)
            for (String row : inputMatrix) {
                // Add zeros at the beginning to extend to desired length
                String extendedRow = "0".repeat(nTarget - cols) + row;
                matrH.add(extendedRow);
            }
            
            // Create generating matrix G of size kTarget x nTarget
            for (int i = 0; i < kTarget; i++) {
                StringBuilder row = new StringBuilder();
                for (int j = 0; j < nTarget; j++) {
                    row.append(j == i ? '1' : '0');  // Identity part
                }
                matrG.add(row.toString());
            }
        } else if ("G".equals(matrixType)) {
            // For matrix G, use the input matrix as is
            matrG.addAll(inputMatrix);
            
            // Create check matrix H of size (n-k) x n
            int n = inputMatrix.get(0).length();
            int k = inputMatrix.size();
            
            for (int i = 0; i < n - k; i++) {
                StringBuilder row = new StringBuilder();
                for (int j = 0; j < n; j++) {
                    row.append(j == k + i ? '1' : '0');  // Identity part
                }
                matrH.add(row.toString());
            }
        }
        
        @SuppressWarnings("unchecked")
        List<String>[] result = new List[2];
        result[0] = matrG;
        result[1] = matrH;
        
        return result;
    }

    /**
     * Настройка блочного кода
     * @param matrix Матрица кода (G или H)
     * @param matrixType Тип матрицы ("G" для порождающей, "H" для проверочной)
     * @return true если настройка успешна, false в противном случае
     */
    public boolean setupCode(List<String> matrix, String matrixType) {
        List<String>[] matrices = generateMatrices(matrix, matrixType);
        if (matrices == null) {
            return false;
        }
        
        List<String> matrG = matrices[0];
        List<String> matrH = matrices[1];
        
        this.k = matrG.size();
        this.n = matrG.get(0).length();
        this.informWords = new ArrayList<>();
        this.codeWords = new ArrayList<>();
        
        // Generate all information words
        for (int i = 0; i < Math.pow(2, this.k); i++) {
            String word = String.format("%" + this.k + "s", Integer.toBinaryString(i)).replace(' ', '0');
            this.informWords.add(word);
        }
        
        // Generate code words
        for (String informWord : this.informWords) {
            List<String> forOperation = new ArrayList<>();
            for (int j = 0; j < informWord.length(); j++) {
                if (informWord.charAt(j) == '1') {
                    forOperation.add(matrG.get(j));
                }
            }
            
            if (!forOperation.isEmpty()) {
                this.codeWords.add(xxor(forOperation));
            } else {
                this.codeWords.add("0".repeat(this.n));
            }
        }
        
        // Determine minimum Hamming distance
        List<Integer> distanceHamm = new ArrayList<>();
        
        // Calculate minimum weight of non-zero code words
        for (String word : this.codeWords) {
            if (!word.equals("0".repeat(this.n))) {  // Skip zero code word
                int weight = 0;
                for (int i = 0; i < word.length(); i++) {
                    if (word.charAt(i) == '1') {
                        weight++;
                    }
                }
                distanceHamm.add(weight);
            }
        }
        
        // For linear code, minimum distance equals minimum weight of non-zero code words
        this.dmin = distanceHamm.isEmpty() ? 0 : distanceHamm.stream().mapToInt(Integer::intValue).min().getAsInt();
        
        // Check for matrix H: if matrix H was passed, check minimum distance by columns of matrix H
        if ("H".equals(matrixType)) {
            // For code with minimum distance 2, each column of H must be non-zero and unique
            // Check columns of original matrix H
            List<String> hColumns = transp(matrix);
            
            // Check that all columns are different and non-zero
            boolean allNonzero = true;
            for (String col : hColumns) {
                if (col.equals("0".repeat(col.length()))) {
                    allNonzero = false;
                    break;
                }
            }
            
            boolean allDistinct = hColumns.stream().distinct().count() == hColumns.size();
            
            if (allNonzero && allDistinct) {
                this.dmin = 2;  // If all columns of matrix H are different and non-zero, then d_min = 2
            }
        }
        
        this.r = this.dmin - 1;  // Number of errors that can be detected
        this.t = (this.dmin - 1) / 2;  // Number of errors that can be corrected
        
        // Generate syndrome table
        this.matrHT = transp(matrH);
        List<String> e = new ArrayList<>();
        List<String> S = new ArrayList<>();
        
        // Error vector with weight 0
        e.add("0".repeat(this.n));
        
        // Error vectors with weight t
        for (int i = 0; i < this.n; i++) {
            if (this.t >= 1) {  // If we can correct at least one error
                // Create vector with one error at position i
                StringBuilder error = new StringBuilder("0".repeat(this.n));
                error.setCharAt(i, '1');
                e.add(error.toString());
            }
        }
        
        // Calculate syndromes
        for (String errorVector : e) {
            List<String> forOperation = new ArrayList<>();
            for (int j = 0; j < errorVector.length(); j++) {
                if (errorVector.charAt(j) == '1') {
                    forOperation.add(this.matrHT.get(j));
                }
            }
            
            if (!forOperation.isEmpty()) {
                S.add(xxor(forOperation));
            } else {
                S.add("0".repeat(this.matrHT.get(0).length()));
            }
        }
        
        // Create syndrome dictionary
        this.sindromVector.clear();
        for (int i = 0; i < e.size(); i++) {
            this.sindromVector.put(S.get(i), e.get(i));
        }
        
        return true;
    }

    /**
     * Кодирование информационного слова
     * @param infoWord Информационное слово
     * @return Закодированное слово
     */
    public String encode(String infoWord) {
        if (this.codeWords.isEmpty() || this.informWords.isEmpty()) {
            return infoWord;
        }
        
        // Pad with zeros to multiple of information word length
        if (infoWord.length() % this.k != 0) {
            infoWord = infoWord + "0".repeat(this.k - (infoWord.length() % this.k));
        }
        
        // Create dictionary for encoding
        Map<String, String> codingDict = new HashMap<>();
        for (int i = 0; i < this.informWords.size(); i++) {
            codingDict.put(this.informWords.get(i), this.codeWords.get(i));
        }
        
        // Encoding
        StringBuilder encodedData = new StringBuilder();
        for (int i = 0; i < infoWord.length(); i += this.k) {
            String informWord = infoWord.substring(i, i + this.k);
            encodedData.append(codingDict.get(informWord));
        }
        
        return encodedData.toString();
    }

    /**
     * Декодирование кодового слова
     * @param receivedCodeWord Полученное кодовое слово
     * @return Декодированное информационное слово
     */
    public String decode(String receivedCodeWord) {
        if (this.codeWords.isEmpty() || this.informWords.isEmpty()) {
            return receivedCodeWord;
        }
        
        // Check if length of encoded data is a multiple of code word length
        if (receivedCodeWord.length() % this.n != 0) {
            return null;
        }
        
        // Decoding
        StringBuilder decodedData = new StringBuilder();
        for (int i = 0; i < receivedCodeWord.length(); i += this.n) {
            String currentCodeWord = receivedCodeWord.substring(i, i + this.n);
            
            // Calculate syndrome
            String syndrome = calculateSyndrome(currentCodeWord);
            
            // Find error pattern
            String errorPattern = sindromVector.get(syndrome);
            if (errorPattern == null) {
                // If syndrome not found, assume no errors
                errorPattern = "0".repeat(this.n);
            }
            
            // Correct errors
            String correctedWord = correctErrors(currentCodeWord, errorPattern);
            
            // Find corresponding information word
            String infoWord = findInformationWord(correctedWord);
            if (infoWord != null) {
                decodedData.append(infoWord);
            }
        }
        
        return decodedData.toString();
    }
    
    // Геттеры для параметров кода
    public int getN() {
        return n;
    }
    
    public int getK() {
        return k;
    }
    
    public int getDmin() {
        return dmin;
    }
    
    public int getR() {
        return r;
    }
    
    public int getT() {
        return this.t;
    }

    public List<String> getCodeWords() {
        return this.codeWords;
    }

    /**
     * Вычисление синдрома для кодового слова
     * @param codeWord Кодовое слово
     * @return Синдром кодового слова
     */
    private String calculateSyndrome(String codeWord) {
        StringBuilder syndrome = new StringBuilder();
        for (int j = 0; j < this.matrHT.get(0).length(); j++) {
            int xorBit = 0;
            for (int k = 0; k < this.n; k++) {
                xorBit ^= Character.getNumericValue(codeWord.charAt(k)) & 
                          Character.getNumericValue(this.matrHT.get(k).charAt(j));
            }
            syndrome.append(xorBit);
        }
        return syndrome.toString();
    }

    /**
     * Исправление ошибок в кодовом слове
     * @param codeWord Кодовое слово с ошибками
     * @param errorPattern Вектор ошибок
     * @return Исправленное кодовое слово
     */
    private String correctErrors(String codeWord, String errorPattern) {
        StringBuilder correctedWord = new StringBuilder();
        for (int j = 0; j < this.n; j++) {
            correctedWord.append(Character.getNumericValue(codeWord.charAt(j)) ^ 
                               Character.getNumericValue(errorPattern.charAt(j)));
        }
        return correctedWord.toString();
    }

    /**
     * Поиск информационного слова по кодовому слову
     * @param codeWord Кодовое слово
     * @return Информационное слово или null, если не найдено
     */
    private String findInformationWord(String codeWord) {
        // Create reverse dictionary for decoding
        Map<String, String> decodingDict = new HashMap<>();
        for (int i = 0; i < this.informWords.size(); i++) {
            decodingDict.put(this.codeWords.get(i), this.informWords.get(i));
        }

        // Try exact match first
        if (decodingDict.containsKey(codeWord)) {
            return decodingDict.get(codeWord);
        }

        // If no exact match, find closest code word
        int minDistance = Integer.MAX_VALUE;
        String bestMatch = null;
        for (String cw : this.codeWords) {
            int distance = 0;
            for (int j = 0; j < codeWord.length(); j++) {
                if (codeWord.charAt(j) != cw.charAt(j)) {
                    distance++;
                }
            }
            if (distance < minDistance) {
                minDistance = distance;
                bestMatch = cw;
            }
        }

        if (bestMatch != null) {
            return decodingDict.get(bestMatch);
        }

        // If we can't find at all, take first k bits
        return codeWord.substring(0, this.k);
    }
} 

File: Theory_Of_Information/6 semester/laboratory__3/src/main/java/task/Лаба 3/block_coder.py
import numpy as np

class BlockCoder:
    def __init__(self):
        self.codeWords = []
        self.informWords = []
        self.matrHT = []
        self.sindromVector = {}
        self.n = 0  # Длина кодового слова
        self.k = 0  # Длина информационного слова
        self.dmin = 0  # Минимальное расстояние кода
        self.r = 0  # Количество ошибок, которые можно обнаружить
        self.t = 0  # Количество ошибок, которые можно исправить

    def xxor(self, array):
        """Логическое XOR для массива строк"""
        if not array:
            return ""
        
        result = [0] * len(array[0])
        for word in array:
            for i in range(len(word)):
                result[i] ^= int(word[i])
        
        return ''.join(str(bit) for bit in result)

    def transp(self, matrix):
        """Транспонирование матрицы"""
        if not matrix:
            return []
        
        result = []
        for i in range(len(matrix[0])):
            new_row = ''.join(row[i] for row in matrix)
            result.append(new_row)
        
        return result

    def generate_matrices(self, input_matrix, matrix_type="G"):
        """Генерация матриц G и H на основе входной матрицы"""
        if not input_matrix or not input_matrix[0]:
            return None, None
            
        if matrix_type == "H":
            # Для специальной матрицы H (4x7) нужно получить n=9, k=5, dmin=2
            rows = len(input_matrix)     # 4 для примера матрицы
            cols = len(input_matrix[0])  # 7 для примера матрицы
            
            # Задаем известные параметры кода n=9, k=5 для матрицы H (4x7)
            n_target = 9  # Желаемая длина кодового слова
            k_target = 5  # Желаемое количество информационных бит
            
            # Создаем матрицу H (расширенная версия входной матрицы)
            matrH = []
            for row in input_matrix:
                # Добавляем нули в начало для расширения до нужной длины
                extended_row = '0' * (n_target - cols) + row
                matrH.append(extended_row)
                
            # Создаем порождающую матрицу G размера k_target x n_target
            matrG = []
            
            # Создаем единичную часть I_k (k x k)
            for i in range(k_target):
                row = ['0'] * n_target
                row[i] = '1'  # Единичная часть
                matrG.append(''.join(row))
            
            return matrG, matrH
            
        elif matrix_type == "G":
            # Преобразование матрицы G работает как прежде
            k = len(input_matrix)  # Количество строк в G (информационные биты)
            n = len(input_matrix[0])  # Количество столбцов в G (кодовое слово)
            
            # Используем исходную матрицу G как есть
            matrG = input_matrix.copy()
            
            # Создаем проверочную матрицу H размера (n-k) x n
            matrH = []
            for i in range(n - k):
                row = ['0'] * n
                row[k + i] = '1'  # Единичная часть
                matrH.append(''.join(row))
            
            return matrG, matrH
        
        else:
            return None, None

    def setup_code(self, matrix, matrix_type="G"):
        """Настройка блочного кода на основе входной матрицы"""
        matrG, matrH = self.generate_matrices(matrix, matrix_type)
        if matrG is None:
            return False

        self.k = len(matrG)
        self.n = len(matrG[0])
        self.informWords = []
        self.codeWords = []

        # Генерация всех информационных слов
        for i in range(2**self.k):
            word = format(i, f'0{self.k}b')
            self.informWords.append(word)

        # Генерация кодовых слов
        for i in range(len(self.informWords)):
            forOperation = []
            for j in range(len(self.informWords[i])):
                if self.informWords[i][j] == "1":
                    forOperation.append(matrG[j])
            
            if forOperation:
                self.codeWords.append(self.xxor(forOperation))
            else:
                self.codeWords.append("0" * self.n)

        # Определение минимального расстояния Хэмминга
        distanceHamm = []
        
        # Вычисляем минимальный вес ненулевых кодовых слов
        for word in self.codeWords:
            if word != "0" * self.n:  # Пропускаем нулевое кодовое слово
                distanceHamm.append(word.count("1"))
        
        # Для линейного кода минимальное расстояние равно минимальному весу ненулевых кодовых слов
        self.dmin = min(distanceHamm) if distanceHamm else 0
        
        # Проверка для матрицы H: если матрица H была передана, проверим минимальное расстояние
        # по столбцам матрицы H
        if matrix_type == "H":
            # Для кода с минимальным расстоянием 2 каждый столбец H должен быть ненулевым и уникальным
            # Проверим столбцы исходной матрицы H
            h_columns = self.transp(matrix)
            
            # Проверяем, что все столбцы разные и ненулевые
            unique_columns = set()
            all_nonzero = True
            
            for col in h_columns:
                if col == "0" * len(col):
                    all_nonzero = False
                    break
                unique_columns.add(col)
            
            all_distinct = len(unique_columns) == len(h_columns)
            
            if all_nonzero and all_distinct:
                self.dmin = 2  # Если все столбцы матрицы H разные и ненулевые, то d_min = 2
        
        self.r = self.dmin - 1  # Количество ошибок, которые можно обнаружить
        self.t = (self.dmin - 1) // 2  # Количество ошибок, которые можно исправить

        # Генерация таблицы синдромов
        self.matrHT = self.transp(matrH)
        self.e = []
        self.S = []
        
        # Вектор ошибок с весом 0
        self.e.append("0" * self.n)
        
        # Векторы ошибок с весом t
        for i in range(self.n):
            if self.t >= 1:  # Если можем исправлять хотя бы одну ошибку
                # Создаем вектор с одной ошибкой в позиции i
                error = ["0"] * self.n
                error[i] = "1"
                self.e.append("".join(error))
        
        # Вычисление синдромов
        for i in range(len(self.e)):
            forOperation = []
            for j in range(len(self.e[i])):
                if self.e[i][j] == "1":
                    forOperation.append(self.matrHT[j])
            
            if forOperation:
                self.S.append(self.xxor(forOperation))
            else:
                self.S.append("0" * len(self.matrHT[0]))
        
        # Создание словаря синдромов
        self.sindromVector = {}
        for i in range(len(self.e)):
            self.sindromVector[self.S[i]] = self.e[i]

        return True

    def encode(self, binary_data):
        """Кодирование бинарных данных с помощью блочного кода"""
        if not self.codeWords or not self.informWords:
            return None

        # Дополнение нулями до кратной длины информационного слова
        if len(binary_data) % self.k != 0:
            binary_data = binary_data + ("0" * (self.k - (len(binary_data) % self.k)))

        # Создание словаря соответствия
        coding_dict = {}
        for i in range(len(self.informWords)):
            coding_dict[self.informWords[i]] = self.codeWords[i]

        # Кодирование
        encoded_data = ''
        for i in range(0, len(binary_data), self.k):
            inform_word = binary_data[i:i+self.k]
            encoded_data += coding_dict[inform_word]

        return encoded_data

    def decode(self, encoded_data):
        """Декодирование с исправлением ошибок"""
        if not self.codeWords or not self.informWords:
            return None

        # Проверка кратности длины закодированных данных
        if len(encoded_data) % self.n != 0:
            return None

        # Создание обратного словаря для декодирования
        decoding_dict = {}
        for i in range(len(self.informWords)):
            decoding_dict[self.codeWords[i]] = self.informWords[i]

        # Декодирование
        decoded_data = ''
        for i in range(0, len(encoded_data), self.n):
            received_word = encoded_data[i:i+self.n]
            
            # Вычисление синдрома
            syndrome = ''
            for j in range(len(self.matrHT[0])):
                xor_bit = 0
                for k in range(self.n):
                    xor_bit ^= int(received_word[k]) & int(self.matrHT[k][j])
                syndrome += str(xor_bit)
            
            # Исправление ошибок, если возможно
            if syndrome in self.sindromVector:
                error_vector = self.sindromVector[syndrome]
                corrected_word = ''
                for j in range(self.n):
                    corrected_word += str(int(received_word[j]) ^ int(error_vector[j]))
                
                # Поиск ближайшего кодового слова
                if corrected_word in decoding_dict:
                    decoded_data += decoding_dict[corrected_word]
                else:
                    # Если не можем найти точное соответствие, берем информационное слово с минимальной дистанцией
                    min_distance = float('inf')
                    best_match = None
                    for code_word in self.codeWords:
                        distance = sum(c1 != c2 for c1, c2 in zip(corrected_word, code_word))
                        if distance < min_distance:
                            min_distance = distance
                            best_match = code_word
                    
                    if best_match:
                        decoded_data += decoding_dict[best_match]
                    else:
                        # Если совсем не можем найти, берем первые k бит
                        decoded_data += corrected_word[:self.k]
            else:
                # Если синдром не найден, ищем ближайшее кодовое слово
                min_distance = float('inf')
                best_match = None
                for code_word in self.codeWords:
                    distance = sum(c1 != c2 for c1, c2 in zip(received_word, code_word))
                    if distance < min_distance:
                        min_distance = distance
                        best_match = code_word
                
                if best_match:
                    decoded_data += decoding_dict[best_match]
                else:
                    # Если совсем не можем найти, берем первые k бит
                    decoded_data += received_word[:self.k]

        return decoded_data 